config/config_test.go

```diff
@@ -70,6 +71,165 @@ func TestDefaultSampler(t *testing.T) {
 	require.Error(t, err)
 }
 
+func TestServiceNameFromEnv(t *testing.T) {
+	os.Setenv(envServiceName, "my-service")
+
+	cfg, err := FromEnv()
+	assert.NoError(t, err)
+
+	_, c, err := cfg.New("")
+	defer c.Close()
+	assert.NoError(t, err)
+	os.Unsetenv(envServiceName)
+}
+
+func TestFromEnv(t *testing.T) {
+	os.Setenv(envServiceName, "my-service")
+	os.Setenv(envDisabled, "false")
+	os.Setenv(envRPCMetrics, "true")
+	os.Setenv(envTags, "KEY=VALUE")
+
+	cfg, err := FromEnv()
+	assert.NoError(t, err)
+	assert.Equal(t, "my-service", cfg.ServiceName)
+	assert.Equal(t, false, cfg.Disabled)
+	assert.Equal(t, true, cfg.RPCMetrics)
+	assert.Equal(t, "KEY", cfg.Tags[0].Key)
+	assert.Equal(t, "VALUE", cfg.Tags[0].Value)
+
+	os.Unsetenv(envServiceName)
+	os.Unsetenv(envDisabled)
+	os.Unsetenv(envRPCMetrics)
+}
+
+func TestNoServiceNameFromEnv(t *testing.T) {
+	os.Unsetenv(envServiceName)
+
+	cfg, err := FromEnv()
+	assert.NoError(t, err)
+
+	_, _, err = cfg.New("")
+	assert.Error(t, err)
+}
+
+func TestSamplerConfigFromEnv(t *testing.T) {
+	// prepare
+	os.Setenv(envSamplerType, "const")
+	os.Setenv(envSamplerParam, "1")
+	os.Setenv(envSamplerManagerHostPort, "http://themaster")
+	os.Setenv(envSamplerMaxOperations, "10")
+	os.Setenv(envSamplerRefreshInterval, "1m1s") // 61 seconds
+
+	// test
+	cfg, err := FromEnv()
+	assert.NoError(t, err)
+
+	// verify
+	assert.Equal(t, "const", cfg.Sampler.Type)
+	assert.Equal(t, float64(1), cfg.Sampler.Param)
+	assert.Equal(t, "http://themaster", cfg.Sampler.SamplingServerURL)
+	assert.Equal(t, int(10), cfg.Sampler.MaxOperations)
+	assert.Equal(t, 61000000000, int(cfg.Sampler.SamplingRefreshInterval))
+
+	// cleanup
+	os.Unsetenv(envSamplerType)
+	os.Unsetenv(envSamplerParam)
+	os.Unsetenv(envSamplerManagerHostPort)
+	os.Unsetenv(envSamplerMaxOperations)
+	os.Unsetenv(envSamplerRefreshInterval)
+}
+
+func TestReporterConfigFromEnv(t *testing.T) {
+	// prepare
+	os.Setenv(envReporterMaxQueueSize, "10")
+	os.Setenv(envReporterFlushInterval, "1m1s") // 61 seconds
+	os.Setenv(envReporterLogSpans, "true")
+	os.Setenv(envAgentHost, "nonlocalhost")
+	os.Setenv(envAgentPort, "6832")
+
+	// test
+	cfg, err := FromEnv()
+	assert.NoError(t, err)
+
+	// verify
+	assert.Equal(t, int(10), cfg.Reporter.QueueSize)
+	assert.Equal(t, 61000000000, int(cfg.Reporter.BufferFlushInterval))
+	assert.Equal(t, true, cfg.Reporter.LogSpans)
+	assert.Equal(t, "nonlocalhost:6832", cfg.Reporter.LocalAgentHostPort)
+
+	// cleanup
+	os.Unsetenv(envReporterMaxQueueSize)
+	os.Unsetenv(envReporterFlushInterval)
+	os.Unsetenv(envReporterLogSpans)
+	os.Unsetenv(envAgentHost)
+	os.Unsetenv(envAgentPort)
+}
+
+func TestParsingErrorsFromEnv(t *testing.T) {
+	os.Setenv(envAgentHost, "localhost") // we require this in order to test the parsing of the port
+
+	tests := []struct {
+		envVar      string
+		value       string
+		expectedErr string
+	}{
+		{
+			envVar:      envRPCMetrics,
+			value:       "NOT_A_BOOLEAN",
+			expectedErr: "strconv.ParseBool",
```

Comment:
see my comment about wrapping errors - if it's implemented, the tests could be written without dependency on the internal wording of Go errors, we can simply be checking for our own strings like `Cannot parse env var JAEGER_RPC_METRICS= NOT_A_BOOLEAN`